bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java

```diff
@@ -217,6 +260,107 @@ protected void afterExecute(Runnable runnable, Throwable throwable) {
+                RejectedExecutionHandler rejectedHandler) {
+            super(corePoolSize, threadFactory, rejectedHandler);
+        }
+
+        @Override
+        protected void beforeExecute(Thread thread, Runnable runnable) {
+            super.beforeExecute(thread, runnable);
+            running.add(runnable);
+        }
+
+        @Override
+        protected void afterExecute(Runnable runnable, Throwable throwable) {
+            super.afterExecute(runnable, throwable);
+            running.remove(runnable);
+            futures.remove(runnable);
+            scheduled.remove(runnable);
+            if (throwable != null) {
+                Throwable cause = throwable.getCause();
+                if (cause instanceof InterruptedException) {
+                    // Ignore this, might happen when we shutdownNow() the executor. We can't
+                    // log at this point as the logging system might be stopped already.
+                    return;
+                }
+            }
+        }
+
+        public void schedule(final Runnable task, final Expression expression) {
+            if (task == null) {
+                throw new NullPointerException();
+            }
+
+            this.setCorePoolSize(this.getCorePoolSize() + 1);
+
+            Runnable scheduleTask = new Runnable() {
+
+                @Override
+                public void run() {
+                    Date now = new Date();
+                    Date time = expression.getTimeAfter(now);
+
+                    try {
+                        while (time != null) {
+                            futures.put(task, ExpressionThreadPoolExecutor.this.schedule(task,
+                                    time.getTime() - now.getTime(), TimeUnit.MILLISECONDS));
+                            scheduled.put(task, this);
+
+                            while (now.before(time)) {
+                                Thread.sleep(time.getTime() - now.getTime());
+                                now = new Date();
+                            }
+
+                            time = expression.getTimeAfter(now);
+                        }
+                    } catch (RejectedExecutionException e) {
+                        logger.error("The executor has already shutdown : '{}'", e.getMessage());
+                    } catch (CancellationException e) {
+                        logger.error("Non executed tasks are cancelled : '{}'", e.getMessage());
+                    } catch (InterruptedException e) {
+                        logger.error("Executing tasks are cancelled : '{}'", e.getMessage());
+                        Thread.currentThread().interrupt();
+                    }
+                }
+            };
+
+            futures.put(scheduleTask, this.submit(scheduleTask));
+        }
+
+        @Override
+        public boolean remove(Runnable task) {
+            Runnable scheduledTask = scheduled.get(task);
```

Comment:
Can't we use <code>Runnable scheduledTask = scheduled.remove(task);</code> here, as it will also return the object if there is one and we cannot loose / forget to remove it if there is an exception or something else?